### Introduction:  
Our aim in this project is to find mobile app profiles for the `App Store` and `Google Play` markets.  
  
We will analyze the apps that are free to download and install, and the main source of revenue for these apps consists of in-app ads. This means that revenue of any given free app is mostly influenced by number of users that use app. Our goal for this project is to analyze data to help to understand what kinds of apps are likely to attract more users.

#### Reading CSV files and separating data from header:

In [1]:
import csv
import pprint as pp

In [2]:
def read_csv(filename):
    with open(filename, encoding='utf8') as fd:
        all_data = list(csv.reader(fd))
        header = all_data[0]
        data = all_data[1:]
    
    return header, data

In [3]:
android_header, android_data = read_csv('googleplaystore.csv')
print('Android data:')
print('Total length: {}'.format(len(android_data)))
for row in android_data[:5]:
    print(row)

Android data:
Total length: 10841
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']
['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']
['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']
['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']
['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


In [4]:
ios_header, ios_data = read_csv('AppleStore.csv')
print('iOS data:')
print('Total length: {}'.format(len(ios_data)))
for row in ios_data[:5]:
    print(row)

iOS data:
Total length: 7197
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']
['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']
['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']
['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']
['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


#### Deleting the wrong data:
From the [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion) of the Google Play dataset, we can see that [this discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) outlines an error for row 10472.

In [5]:
print(android_data[10472])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [6]:
del(android_data[10472])
print(android_data[10472])

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


#### Removing duplicate entries:
As we explore Google Play dataset further, we will find some apps that have more than one entry. For example, `Instagram` has 4 entries.

In [7]:
for row in android_data:
    if row[0] == 'Instagram':
        print(row)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


In [8]:
def check_duplicates(dataset, name_ind):
    unique_apps = set()
    
    for row in dataset:
        app = row[name_ind]
        unique_apps.add(app)
    
    return len(dataset) - len(unique_apps)

In [9]:
print('Number of duplicates in Google Play dataset: {}'.format(check_duplicates(android_data, 0))) # column 1, 'App'

Number of duplicates in Google Play dataset: 1181


In [10]:
print('Number of duplicates in App Store dataset: {}'.format(check_duplicates(ios_data, 1))) # column 2, 'track_name'

Number of duplicates in App Store dataset: 2


In total there are `1181` duplicate app entries in Google Play dataset.  
We don't want to count certain apps more than once when we analyze data, so we need to remove the duplicate entries and keep only one entry per app.  
  
When we examine duplicate entries of `Instagram`, the main difference happens on 4th column of each row (`Reviews`), which corresponds to the number of reviews. The different numtbers show that the data was collected at different times. We can use this to build a criterion for keeping rows. We will keep the rows that have the highest number of reviews because the higher the number of reviews, the more reliable the ratings.  
  
To remove the duplicate entries, we will:
* Create a dictionary where each key is an unique app name, and the value is the highest number of reviews of that app.
* Use the above created dictionary to create a new dataset, which will have only one entry per app.

In [11]:
def get_max_reviews(dataset, name_ind, review_ind):
    reviews_max = dict()
    
    for row in dataset:
        app = row[name_ind]
        review = float(row[review_ind])

        if app in reviews_max and reviews_max[app] < review:
            reviews_max[app] = review
        elif app not in reviews_max:
            reviews_max[app] = review
    
    return reviews_max

In [12]:
def remove_duplicate_apps(dataset, name_ind, review_ind):
    
    if check_duplicates(dataset, name_ind) > 0:
        reviews_max = get_max_reviews(dataset, name_ind, review_ind)

        clean_data = list()
        already_added = list()
    
        for row in dataset:
            app = row[name_ind]
            review = float(row[review_ind])
        
            if reviews_max[app] == review and (app not in already_added):
                clean_data.append(row)
                already_added.append(app)
    
        return clean_data
    else:
        return dataset

In [13]:
android_data = remove_duplicate_apps(android_data, 0, 3) # columns 'App' and 'Review'
print('Android data:')
print('Length after removing duplicates: {}'.format(len(android_data)))
for row in android_data[:5]:
    print(row)

Android data:
Length after removing duplicates: 9659
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']
['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']
['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']
['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']
['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


In [14]:
ios_data = remove_duplicate_apps(ios_data, 1, 5) # columns 'track_name', 'rating_count_tot'
print('iOS data:')
print('Length after removing duplicates: {}'.format(len(ios_data)))
for row in ios_data[:5]:
    print(row)

iOS data:
Length after removing duplicates: 7195
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']
['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']
['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']
['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']
['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


#### Removing non-English apps:  
As we explore the datasets even further, we can notice the names of the some of the app suggest that they are not directed towards an English-speaking audience. Below are the example:

In [15]:
print(android_data[4412][0])
print(android_data[7940][0])

中国語 AQリスニング
لعبة تقدر تربح DZ


We are not interested in keeping these kind of apps, so we will remove them. One way is to, remove each app whose name contains a symbol that is not commonly used in English text - English text usually includes the letters from English alphabet (a-zA-Z), numbers (0-9), punctuations (.,;?! etc), and other symbols (+-* / etc).  
  
All these characters that are specific to English texts are encoded using the ASCII standards. Each ASCII character has a corresponding number between 0-127 associated with it, and we can take advantage of that to build a function that checks an app name and tell us whether it contains non-ASCII characters.  
But some English app name use emojis or other symbols (™, 😜 etc) that fall outside of the ASCII range. To minimize the impact of loss, we will only remove an app if its name has more than 3 non-ASCII characters.

In [16]:
def is_english(app_name):
    non_ascii = 0
    
    for char in app_name:
        if ord(char) > 127:
            non_ascii += 1
    
    return (non_ascii <= 3)

In [17]:
def english_dataset(dataset, name_ind):
    clean_data = list()
    
    for row in dataset:
        app = row[name_ind]
        if is_english(app):
            clean_data.append(row)
    
    return clean_data

In [18]:
android_data = english_dataset(android_data, 0) # column 'App'
print('Android data:')
print('Length after removing non-English apps: {}'.format(len(android_data)))
for row in android_data[:5]:
    print(row)

Android data:
Length after removing non-English apps: 9614
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']
['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']
['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']
['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']
['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


In [19]:
ios_data = english_dataset(ios_data, 1) # column 'track_name'
print('iOS data:')
print('Length after removing non-English apps: {}'.format(len(ios_data)))
for row in ios_data[:5]:
    print(row)

iOS data:
Length after removing non-English apps: 6181
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']
['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']
['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']
['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']
['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


#### Isolating the free apps:  
As mentioned in the introduction, we will analyze the apps that are free to download and install, and revenue of free apps are in-app ads. Since our dataset contains both free and non-free apps, we will need to isolate only the free apps for our analysis.

In [20]:
def get_free_apps(dataset, price_ind):
    free_apps = list()
    
    for row in dataset:
        price = row[price_ind].replace('$', '')
        if float(price) == 0.0:
            free_apps.append(row)
    
    return free_apps

In [21]:
android_data = get_free_apps(android_data, 7) # column 'Price'
print('Android data:')
print('Length after removing non-free apps: {}'.format(len(android_data)))
for row in android_data[:5]:
    print(row)

Android data:
Length after removing non-free apps: 8864
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']
['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']
['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']
['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']
['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


In [22]:
ios_data = get_free_apps(ios_data, 4) # column 'price'
print('iOS data:')
print('Length after removing non-free apps: {}'.format(len(ios_data)))
for row in ios_data[:5]:
    print(row)

iOS data:
Length after removing non-free apps: 3220
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']
['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']
['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']
['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']
['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


We are now left with `8864` android apps and `3220` iOS apps.  
  
#### Most common apps by genre:  
Let us begin our analysis by getting a sense of the most common genres for each market. For this, we will build a frequency table.  
  
We will build two functions which we can use to analyze the frequency tables:
* One function to generate frequency table that shows percentage
* Another function that we can use to display the percentages in descending order

In [23]:
def frequency_table(dataset, genre_ind):
    freq_table = dict()
    
    for row in dataset:
        genre = row[genre_ind]
        freq_table[genre] = freq_table.get(genre, 0) + 1
    
    total = len(dataset)
    percent_table = dict()
    for genre in freq_table:
        percent_table[genre] = (freq_table[genre] / total) * 100
    
    return percent_table

In [24]:
def display_table(freq_table):
    # sort in descending order
    table_sorted = sorted(freq_table.items(), key=lambda x: x[1], reverse=True)
    for entry in table_sorted:
        print('{} : {:.2f}'.format(*entry))

Examining frequency table for the `prime_genre` column of the App Store data set.

In [25]:
prime_genre_freq = frequency_table(ios_data, -5) # column 'prime_genre'
display_table(prime_genre_freq)

Games : 58.14
Entertainment : 7.89
Photo & Video : 4.97
Education : 3.66
Social Networking : 3.29
Shopping : 2.61
Utilities : 2.52
Sports : 2.14
Music : 2.05
Health & Fitness : 2.02
Productivity : 1.74
Lifestyle : 1.58
News : 1.34
Travel : 1.24
Finance : 1.12
Weather : 0.87
Food & Drink : 0.81
Reference : 0.56
Business : 0.53
Book : 0.43
Navigation : 0.19
Medical : 0.19
Catalogs : 0.12


We can see that among the free English apps, more than half (`58.14%`) are `Games`. `Entertainment` apps are close to `8%`, followed by `Photo & Video` apps, which is close to `5%`. Only `3.66%` of the apps are designed for `Education`, followed by `Social Networking` apps which amount `3.29%` of the apps in our dataset.  
  
The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare.  
  
Examining the `Category` and `Genre` columns of the Google Play dataset:

In [26]:
category_freq = frequency_table(android_data, 1) # column 'Category'
display_table(category_freq)

FAMILY : 18.91
GAME : 9.72
TOOLS : 8.46
BUSINESS : 4.59
LIFESTYLE : 3.90
PRODUCTIVITY : 3.89
FINANCE : 3.70
MEDICAL : 3.53
SPORTS : 3.40
PERSONALIZATION : 3.32
COMMUNICATION : 3.24
HEALTH_AND_FITNESS : 3.08
PHOTOGRAPHY : 2.94
NEWS_AND_MAGAZINES : 2.80
SOCIAL : 2.66
TRAVEL_AND_LOCAL : 2.34
SHOPPING : 2.25
BOOKS_AND_REFERENCE : 2.14
DATING : 1.86
VIDEO_PLAYERS : 1.79
MAPS_AND_NAVIGATION : 1.40
FOOD_AND_DRINK : 1.24
EDUCATION : 1.16
ENTERTAINMENT : 0.96
LIBRARIES_AND_DEMO : 0.94
AUTO_AND_VEHICLES : 0.93
HOUSE_AND_HOME : 0.82
WEATHER : 0.80
EVENTS : 0.71
PARENTING : 0.65
ART_AND_DESIGN : 0.64
COMICS : 0.62
BEAUTY : 0.60


There are not many apps designed for fun in Google play, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle etc). However, if we investigate this further, we can see that the family category means mostly games for kids.

In [27]:
genre_freq = frequency_table(android_data, -4) # column 'Genre'
display_table(genre_freq)

Tools : 8.45
Entertainment : 6.07
Education : 5.35
Business : 4.59
Lifestyle : 3.89
Productivity : 3.89
Finance : 3.70
Medical : 3.53
Sports : 3.46
Personalization : 3.32
Communication : 3.24
Action : 3.10
Health & Fitness : 3.08
Photography : 2.94
News & Magazines : 2.80
Social : 2.66
Travel & Local : 2.32
Shopping : 2.25
Books & Reference : 2.14
Simulation : 2.04
Dating : 1.86
Arcade : 1.85
Video Players & Editors : 1.77
Casual : 1.76
Maps & Navigation : 1.40
Food & Drink : 1.24
Puzzle : 1.13
Racing : 0.99
Libraries & Demo : 0.94
Role Playing : 0.94
Auto & Vehicles : 0.93
Strategy : 0.91
House & Home : 0.82
Weather : 0.80
Events : 0.71
Adventure : 0.68
Comics : 0.61
Art & Design : 0.60
Beauty : 0.60
Parenting : 0.50
Card : 0.45
Casino : 0.43
Trivia : 0.42
Educational;Education : 0.39
Board : 0.38
Educational : 0.37
Education;Education : 0.34
Word : 0.26
Casual;Pretend Play : 0.24
Music : 0.20
Entertainment;Music & Video : 0.17
Puzzle;Brain Games : 0.17
Racing;Action & Adventure : 0.1

The difference between the *Genres* and the *Category* columns is not clear, but one thing that we can notice is that the *Genres* column has more categories.  
Up to this point, we found that the AppStore is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps.  
  
#### Most popular apps by Genre on the App Store:  
One way to find out what genres are the most popular is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the `Installs` column, but for the App Store data, this information is missing. As a workaround, we shall take the total number of user ratings as a proxy, which we can find in the `rating_count_tot` app.  
  
Below, we shall calculate the average number of user ratings per app genre on the App Store:

In [28]:
def rating_per_genre(dataset, freq_table):
    genre_ratings = dict()
    
    for genre in freq_table:
        genre_len = 0
        ratings = 0
        
        for row in dataset:
            app_genre = row[-5]
            if app_genre == genre:
                ratings += float(row[5])
                genre_len += 1
        avg_ratings = ratings / genre_len
        genre_ratings[genre] = avg_ratings
    
    return genre_ratings

In [29]:
genre_ios = frequency_table(ios_data, -5)
ios_genre_ratings = rating_per_genre(ios_data, genre_ios)
display_table(ios_genre_ratings)

Navigation : 86090.33
Reference : 74942.11
Social Networking : 71548.35
Music : 57326.53
Weather : 52279.89
Book : 39758.50
Food & Drink : 33333.92
Finance : 31467.94
Photo & Video : 28441.54
Travel : 28243.80
Shopping : 26919.69
Health & Fitness : 23298.02
Sports : 23008.90
Games : 22812.92
News : 21248.02
Productivity : 21028.41
Utilities : 18684.46
Lifestyle : 16485.76
Entertainment : 14029.83
Business : 7491.12
Education : 7003.98
Catalogs : 4004.00
Medical : 612.00


In [30]:
for row in ios_data:
    if row[-5] == 'Navigation':
        print('{0[1]}: {0[5]}'.format(row))

Waze - GPS Navigation, Maps & Real-time Traffic: 345046
Google Maps - Navigation & Transit: 154911
Geocaching®: 12811
CoPilot GPS – Car Navigation & Offline Maps: 3582
ImmobilienScout24: Real Estate Search in Germany: 187
Railway Route Search: 5


On average, `Navigation` apps have the highest number of user reviews. But this figure is heavily influenced by `Waze` and `Google Maps`, which as around `500000` user reviews together.  
  
The same pattern applies to `Social Networking` apps, where average user reviews is heavily influenced by popular apps like `Facebook`, `Instagram`, `Pinterest`, `Skype` etc. Same applies to music apps, where `Pandora`, `Spotify`, `Shazam` heavily influence average ratings.  
  
Our aim is to find popular genres, but `Navigation`, `Social Networking` or `Music` apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages.

In [31]:
for row in ios_data:
    if row[-5] == 'Reference':
        print('{0[1]}: {0[5]}'.format(row))

Bible: 985920
Dictionary.com Dictionary & Thesaurus: 200047
Dictionary.com Dictionary & Thesaurus for iPad: 54175
Google Translate: 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran: 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition: 17588
Merriam-Webster Dictionary: 16849
Night Sky: 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE): 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools: 4693
GUNS MODS for Minecraft PC Edition - Mods Tools: 1497
Guides for Pokémon GO - Pokemon GO News and Cheats: 826
WWDC: 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free: 718
VPN Express: 14
Real Bike Traffic Rider Virtual Reality Glasses: 8
教えて!goo: 0
Jishokun-Japanese English Dictionary & Translator: 0


`Reference` apps has around `75000` average user ratings, but it is actually the `Bible` and `Dictionary.com` which skew up the user ratings.  
  
However, this niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.  
  
This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.  
  
Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:
* `Weather` apps: People generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.
* `Food & Drink` apps: Examples here include `Starbucks`, `Dunkin' Donuts`, `McDonald's`, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope.
* `Finance` apps: These apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge.  
  
Now let's analyze the Google Play market.  
  
#### Most popular apps by genre on Google Play:  
For the Google Play market, we actually have data about the number of installs, so we will be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough, we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [32]:
install_freq = frequency_table(android_data, 5)
display_table(install_freq)

1,000,000+ : 15.73
100,000+ : 11.55
10,000,000+ : 10.55
10,000+ : 10.20
1,000+ : 8.39
100+ : 6.92
5,000,000+ : 6.83
500,000+ : 5.56
50,000+ : 4.77
5,000+ : 4.51
10+ : 3.54
500+ : 3.25
50,000,000+ : 2.30
100,000,000+ : 2.13
50+ : 1.92
5+ : 0.79
1+ : 0.51
500,000,000+ : 0.27
1,000,000,000+ : 0.23
0+ : 0.05
0 : 0.01


One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes, we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.  
  
We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.  
  
To perform computations, however,
* We'll need to convert each install number to float.
* This means that we need to remove the commas and the plus characters.  
We also compute the average number of installs for each genre (`Category`).

In [33]:
def installs_per_genre(dataset, freq_table):
    avg_installs = dict()
    
    for category in freq_table:
        total_install = 0
        category_len = 0
        
        for row in dataset:
            app_category = row[1]
            if app_category == category:
                install = row[5]
                for char in [',', '+']:
                    install = install.replace(char, '')
                total_install += float(install)
                category_len += 1
        avg = total_install/category_len
        avg_installs[category] = avg
    
    return avg_installs

In [34]:
category_freq = frequency_table(android_data, 1) # column 'Category'
android_install = installs_per_genre(android_data, category_freq)
display_table(android_install)

COMMUNICATION : 38456119.17
VIDEO_PLAYERS : 24727872.45
SOCIAL : 23253652.13
PHOTOGRAPHY : 17840110.40
PRODUCTIVITY : 16787331.34
GAME : 15588015.60
TRAVEL_AND_LOCAL : 13984077.71
ENTERTAINMENT : 11640705.88
TOOLS : 10801391.30
NEWS_AND_MAGAZINES : 9549178.47
BOOKS_AND_REFERENCE : 8767811.89
SHOPPING : 7036877.31
PERSONALIZATION : 5201482.61
WEATHER : 5074486.20
HEALTH_AND_FITNESS : 4188821.99
MAPS_AND_NAVIGATION : 4056941.77
FAMILY : 3695641.82
SPORTS : 3638640.14
ART_AND_DESIGN : 1986335.09
FOOD_AND_DRINK : 1924897.74
EDUCATION : 1833495.15
BUSINESS : 1712290.15
LIFESTYLE : 1437816.27
FINANCE : 1387692.48
HOUSE_AND_HOME : 1331540.56
DATING : 854028.83
COMICS : 817657.27
AUTO_AND_VEHICLES : 647317.82
LIBRARIES_AND_DEMO : 638503.73
PARENTING : 542603.62
BEAUTY : 513151.89
EVENTS : 253542.22
MEDICAL : 120550.62


In [35]:
for row in android_data:
    if row[1] == 'COMMUNICATION' and (row[5] in ['1,000,000,000+', '500,000,000+', '100,000,000+']):
        print('{0[0]}: {0[5]}'.format(row))

WhatsApp Messenger: 1,000,000,000+
imo beta free calls and text: 100,000,000+
Android Messages: 100,000,000+
Google Duo - High Quality Video Calls: 500,000,000+
Messenger – Text and Video Chat for Free: 1,000,000,000+
imo free video calls and chat: 500,000,000+
Skype - free IM & video calls: 1,000,000,000+
Who: 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji: 100,000,000+
LINE: Free Calls & Messages: 500,000,000+
Google Chrome: Fast & Secure: 1,000,000,000+
Firefox Browser fast & private: 100,000,000+
UC Browser - Fast Download Private & Secure: 500,000,000+
Gmail: 1,000,000,000+
Hangouts: 1,000,000,000+
Messenger Lite: Free Calls & Messages: 100,000,000+
Kik: 100,000,000+
KakaoTalk: Free Calls & Text: 100,000,000+
Opera Mini - fast web browser: 100,000,000+
Opera Browser: Fast and Secure: 100,000,000+
Telegram: 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer: 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure: 100,000,000+
Viber Messenger: 500,000,000+
WeC

On average, communication apps have more installs: `38456119`. This number is heavily skewed up by a dew apps that have over `1,000,000` installs like `WhatsApp`, `Facebook Messenger`, `Skype`, `Google Chrome`, `Gmail`, `Hangout` and few others with over `500,000,000` and `100,000,000` installs.